File: rust/src/web.rs
//! Trình duyệt của huba — lái bằng Playwright, qua `web.mjs`.
//!
//! 🔴 Hà 2026-08-23: *"Sao khong dùng playwright"*, ngay sau *"Tôi ko ngồi
//! máy"*. Đường AppleScript ([`crate::browser`]) lái đúng cái Chrome đang đăng
//! nhập của chủ máy, nhưng nó cần quyền Tự động hoá `hubd → Google Chrome` —
//! **cấp quyền ấy phải ngồi trước máy**. Một cây cầu mà nhịp đầu chỉ bắc được
//! khi đang đứng ở bờ bên kia thì chưa phải cầu.
//!
//! CDP không đi qua Apple Events, cũng không cần Screen Recording: **không
//! quyền macOS nào**, chạy được cả khi màn hình khoá. Số đo và ba chỗ đã trả
//! giá nằm ở đầu `web.mjs`; ở đây chỉ giữ phần Rust phải biết.

use std::path::{Path, PathBuf};
use std::time::Duration;

use serde_json::{json, Value};

use crate::exec::{run, RunOpts};
use crate::logging;

/// Trần cho một lệnh trình duyệt. Rộng hơn `osascript` nhiều vì lượt ĐẦU phải
/// dựng cả một trình duyệt, và một trang chậm còn được `web.mjs` chờ tới 30s.
const TIMEOUT: Duration = Duration::from_secs(75);

/// 🔴 ĐƯỜNG TUYỆT ĐỐI, KHÔNG PHẢI `node`. Job của launchd chạy với PATH tối
/// thiểu (`/usr/bin:/bin:/usr/sbin:/sbin`), nên gọi trống tên là `rc=127` —
/// bài học đã ghi trong sổ workspace, và nó im lặng theo kiểu tệ nhất: lệnh
/// "chạy xong" mà chẳng có gì xảy ra.
fn node_bin() -> Option<PathBuf> {
    let mut ung = Vec::new();
    if let Ok(v) = std::env::var("HUB_NODE") {
        ung.push(PathBuf::from(v));
    }
    ung.push(PathBuf::from("/usr/local/bin/node"));
    ung.push(PathBuf::from("/opt/homebrew/bin/node"));
    ung.push(PathBuf::from("/usr/bin/node"));
    ung.into_iter().find(|p| p.is_file())
}

/// Gọi `web.mjs` một lượt và đọc dòng JSON nó in ra.
///
/// Mọi lỗi ra ngoài dưới dạng `Err(String)` đã thành CÂU cho người đọc: chỗ gọi
/// là một tin nhắn Telegram, không phải một stack trace.
pub fn call(hub_home: &Path, lenh: &Value) -> Result<Value, String> {
    let script = hub_home.join("web.mjs");
    if !script.is_file() {
        return Err(format!("không thấy {}", script.display()));
    }
    let node = node_bin().ok_or_else(|| {
        "không thấy `node` trên máy (đã tìm HUB_NODE · /usr/local/bin · /opt/homebrew/bin). \
         Trình duyệt của huba cần Node để chạy Playwright."
            .to_string()
    })?;
    let arg = lenh.to_string();
    let out = run(
        &node.display().to_string(),
        &[&script.display().to_string(), &arg],
        RunOpts {
            cwd: Some(hub_home),
            timeout: Some(TIMEOUT),
            ..Default::default()
        },
    )
    .map_err(|e| e.to_string())?;
    if out.timed_out {
        return Err(format!(
            "lệnh trình duyệt quá {}s — xem logs/browser.log",
            TIMEOUT.as_secs()
        ));
    }
    // 🔴 ĐỌC DÒNG JSON, KHÔNG ĐỌC MÃ THOÁT. `web.mjs` trả mã 1 kèm một câu lỗi
    // ĐÃ VIẾT SẴN cho người đọc; vứt câu ấy đi để in "exit 1" là đúng thứ đã
    // chặn mọi bản vá của huba hôm 22/08 (`exec::drain_capped`).
    let dong = out
        .stdout
        .lines()
        .rev()
        .find(|l| l.trim_start().starts_with('{'))
        .unwrap_or("");
    let v: Value = serde_json::from_str(dong).unwrap_or_else(|_| json!({}));
    if let Some(loi) = v.get("error").and_then(Value::as_str) {
        logging::warn(
            "web_cmd_failed",
            json!({ "do": lenh.get("do"), "err": loi }),
        );
        return Err(loi.to_string());
    }
    if !out.ok() {
        // Mã thoát khác 0 mà KHÔNG có dòng JSON nào: node chết trước khi kịp
        // nói. Lúc ấy `stderr` là thứ duy nhất còn lại — đừng nuốt nó.
        return Err(format!(
            "trình duyệt không trả lời (mã {:?}): {}",
            out.code,
            crate::exec::truncate(out.stderr.trim(), 200)
        ));
    }
    Ok(v)
}

/// Một câu ngắn nói trang đang mở là trang nào.
fn noi_trang(v: &Value) -> String {
    let title = v.get("title").and_then(Value::as_str).unwrap_or("").trim();
    let url = v.get("url").and_then(Value::as_str).unwrap_or("");
    let host = crate::pipeline::web_host(url);
    if title.is_empty() {
        host
    } else {
        format!("{title} · {host}")
    }
}

/// Trần chữ cho một trang gửi về Telegram. Cắt thì phải NÓI: một trang cắt im
/// lặng đọc lên y hệt một trang ngắn.
const TEXT_MAX: usize = 3500;

/// Trình duyệt ẨN của huba có đang chạy không — và pid của nó.
///
/// 🔴 Có mặt vì một phép đo 2026-08-23: khi bản ẩn còn sống, **Apple Events
/// trỏ vào NÓ chứ không vào Chrome của chủ máy**. `browser::tabs()` đọc ra
/// đúng một tab `iana.org` — trang của bản ẩn — rồi giết bản ẩn đi thì Chrome
/// thật trả về `0` cửa sổ. Hai thế giới cùng một bundle id, và Apple Events
/// không phân biệt được. Nên đường "Chrome thật" phải HỎI trước rồi mới nói,
/// không thì nó khai một điều sai về thế giới — thứ tệ hơn hẳn im lặng.
pub fn an_dang_chay(hub_home: &Path) -> Option<u32> {
    let pid: u32 = std::fs::read_to_string(hub_home.join("data/browser.pid"))
        .ok()?
        .trim()
        .parse()
        .ok()?;
    // `kill -0` chỉ HỎI, không giết.
    con_song(pid).then_some(pid)
}

/// Tiến trình ấy còn sống không — hỏi bằng `kill -0`, thứ chỉ HỎI chứ không
/// giết.
///
/// Đi bằng một tiến trình con chứ không kéo `libc` vào: `unsafe` của repo này
/// sống ở ĐÚNG MỘT TỆP (`cgkeys.rs`, xem `CLAUDE.md`), và một lời gọi `kill(2)`
/// trực tiếp sẽ mở cái cửa ấy ra ở tệp thứ hai để đổi lấy vài micro giây.
fn con_song(pid: u32) -> bool {
    crate::exec::run(
        "kill",
        &["-0", &pid.to_string()],
        RunOpts {
            timeout: Some(Duration::from_secs(5)),
            ..Default::default()
        },
    )
    .map(|o| o.ok())
    .unwrap_or(false)
}

/// `/web` — một chỗ quyết định cho mọi dạng tham số.
///
/// Trả `(câu trả lời, đường dẫn ảnh nếu có)`.
pub fn route(hub_home: &Path, want: &str) -> (String, Option<PathBuf>) {
    let want = want.trim();
    let lenh = match want {
        "" => json!({"do":"where"}),
        "doc" | "đọc" | "text" | "chu" | "chữ" => json!({"do":"text"}),
        "tat" | "tắt" | "close" => json!({"do":"close"}),
        "enter" => json!({"do":"press","key":"Enter"}),
        _ => {
            if let Some(chu) = want
                .strip_prefix("bấm ")
                .or_else(|| want.strip_prefix("bam "))
                .or_else(|| want.strip_prefix("click "))
            {
                json!({"do":"click","what": chu.trim()})
            } else if let Some(chu) = want
                .strip_prefix("gõ ")
                .or_else(|| want.strip_prefix("go "))
                .or_else(|| want.strip_prefix("type "))
            {
                // KHÔNG `trim()` phần chữ: khoảng trắng đầu/cuối có thể là một
                // phần của thứ người ta muốn gõ, và ở đây không có cách nào
                // đoán hộ cho đúng.
                json!({"do":"type","text": chu})
            } else if let Some(o) = want
                .strip_prefix("ô ")
                .or_else(|| want.strip_prefix("o "))
                .or_else(|| want.strip_prefix("field "))
            {
                json!({"do":"field","what": o.trim()})
            } else {
                match crate::browser::dia_chi_hop_le(want) {
                    Some(url) => json!({"do":"goto","url": url}),
                    None => {
                        return (
                            "Không hiểu. `/web an` = đang ở đâu · `… <địa chỉ>` = mở · \
                             `… doc` = đọc chữ · `… bấm <chữ>` · `… ô <nhãn>` = chọn ô nhập · \
                             `… gõ <chữ>` · `… enter` · `… tắt`"
                                .to_string(),
                            None,
                        )
                    }
                }
            }
        }
    };
    let v = match call(hub_home, &lenh) {
        Ok(v) => v,
        Err(e) => return (format!("⚠ {e}"), None),
    };
    if lenh["do"] == json!("close") {
        return ("🌐 Đã tắt trình duyệt của huba.".to_string(), None);
    }
    if lenh["do"] == json!("text") {
        let chu = v.get("text").and_then(Value::as_str).unwrap_or("").trim();
        if chu.is_empty() {
            return (
                format!("{} — trang này không có chữ nào.", noi_trang(&v)),
                None,
            );
        }
        let n = chu.chars().count();
        let than = if n > TEXT_MAX {
            format!(
                "{}\n\n… cắt ở {TEXT_MAX} ký tự (trang dài {n}).",
                chu.chars().take(TEXT_MAX).collect::<String>()
            )
        } else {
            chu.to_string()
        };
        return (format!("🌐 {}\n\n{than}", noi_trang(&v)), None);
    }
    // Còn lại đều là "đang đứng ở một trang": kèm ẢNH, vì thứ người cầm điện
    // thoại cần trước hết là NHÌN THẤY trang, rồi mới tới chữ.
    let anh = hub_home.join("data").join("web-shot.png");
    let co_anh = call(
        hub_home,
        &json!({"do":"shot","path": anh.display().to_string()}),
    )
    .is_ok()
        && anh.is_file();
    (format!("🌐 {}", noi_trang(&v)), co_anh.then_some(anh))
}

File: web.mjs
// Bộ lái trình duyệt của huba — MỘT lệnh, một dòng JSON ra.
//
// 🔴 Hà 2026-08-23: *"Sao khong dùng playwright"*, hỏi ngay sau *"Tôi ko ngồi
// máy"*. Hai câu ấy là một câu, và nó lật lựa chọn đầu của tôi:
//
// * Đường AppleScript (`rust/src/browser.rs`) lái đúng cái Chrome đang đăng
//   nhập của chủ máy — nghe hợp phép thử cầu nối hơn. Nhưng nó cần quyền Tự
//   động hoá `hubd → Google Chrome`, mà **cấp quyền ấy phải ngồi trước máy**.
//   Một tính năng cho người ở xa mà cánh cửa đầu tiên của nó chỉ mở được khi
//   ngồi ở nhà thì chưa phải là cầu.
// * CDP không đi qua Apple Events, cũng không cần Screen Recording — **không
//   quyền macOS nào**. Nên đường này chạy được ngay bây giờ, từ điện thoại, và
//   chạy cả khi màn hình đang khoá.
//
// ## Ba số đo quyết định hình dạng tệp này, đo trên chính máy này 2026-08-23
//
// 1. **Đường "không cần Node" đã thử rồi mới bỏ.** `Google Chrome --headless
//    --dump-dom https://example.com` in ra đúng 561 byte DOM thật — nhưng tiến
//    trình KHÔNG thoát, chạy 3 phút vẫn treo, phải giết. Muốn dùng thì phải tự
//    đọc stdout rồi tự giết: đúng thứ `CLAUDE.md` gọi là đọc mã thoát của một
//    thứ chỉ khởi chạy. Playwright sinh ra để nuốt đúng lớp đó.
// 2. **Binary của Playwright KHÔNG ra được mạng trên máy này.** Cùng một đường
//    CDP, cùng một đoạn mã: `Chrome for Testing` (chromium-1228 trong cache
//    ms-playwright) treo hết 20 giây ở `page.goto`; `/Applications/Google
//    Chrome.app` xong trong **169 ms**. Biến số duy nhất là cái binary — nên
//    huba lái **Chrome thật**, và Playwright chỉ làm bộ lái.
// 3. **Hồ sơ mặc định không mở CDP được** (Chrome cấm), nên phải có hồ sơ
//    riêng. Đó vừa là giới hạn vừa là tính năng: không thừa hưởng phiên đăng
//    nhập nào của Chrome thật, nhưng GIỮ được phiên nó tự tạo — đăng nhập một
//    lần từ điện thoại thì lần sau còn nguyên.

import { chromium } from 'playwright'
import { spawn } from 'node:child_process'
import { mkdirSync, existsSync, readFileSync, writeFileSync, openSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

const HERE = dirname(fileURLToPath(import.meta.url))
const PORT = Number(process.env.HUB_WEB_PORT ?? 9334)
const PROFILE = join(HERE, 'data', 'browser-profile')
const LOCK = join(HERE, 'data', 'browser.pid')
// 🔴 Hà 2026-08-23: *"Làm sao giữ được cache, vì tôi thấy mỗi lần mở lại mất hết
// trạng thái cũ"*. Đo rồi mới vá, và hoá ra hai vế khác nhau:
//
// * **Cache thì KHÔNG mất.** Đặt `localStorage` + cookie, tắt hẳn trình duyệt,
//   mở lại — cả hai còn nguyên (hồ sơ đang giữ 112 MB thật). Đăng nhập một lần
//   là còn mãi; đó là công của `--user-data-dir`.
// * Thứ mất là **đang đứng ở trang nào**: Chrome khởi động về `about:blank`.
//   Với người cầm điện thoại thì đó đúng là "mất trạng thái" — mở lên thấy
//   trắng trơn, phải gõ lại địa chỉ.
//
// Nên nhớ lấy địa chỉ cuối, và lượt dựng NGUỘI thì tự quay về đó.
const LAST = join(HERE, 'data', 'browser-last.json')
const LOG = join(HERE, 'logs', 'browser.log')
const CONNECT_TIMEOUT_MS = 20_000

/// Chrome THẬT, không phải bản Playwright tải về — xem số đo 2 ở đầu tệp.
const BINS = [
  process.env.HUB_BROWSER_BIN,
  '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
  '/Applications/Brave Browser.app/Contents/MacOS/Brave Browser',
].filter(Boolean)

function say(o) {
  process.stdout.write(JSON.stringify(o) + '\n')
}
function die(msg) {
  say({ error: String(msg) })
  process.exit(1)
}

async function noiVao(ms = 3000) {
  try {
    return await chromium.connectOverCDP(`http://127.0.0.1:${PORT}`, { timeout: ms })
  } catch {
    return null
  }
}

function conDangSong() {
  try {
    const pid = Number(readFileSync(LOCK, 'utf8').trim())
    if (!pid) return 0
    process.kill(pid, 0) // chỉ HỎI, không giết
    return pid
  } catch {
    return 0
  }
}

async function dungTrinhDuyet() {
  // 🔴 MỘT BẢN, KHÔNG HAI. Bản đầu của tệp này dựng thêm một chromium mỗi khi
  // nối hụt, và ba tiến trình cùng giành một hồ sơ thì **cả ba treo** — đo
  // 23/08: cổng vẫn LISTEN mà `connectOverCDP` hết giờ ở 10 giây. Nên hỏi cái
  // khoá trước: có bản đang sống thì CHỜ nó, đừng dựng bản nữa.
  const cu = conDangSong()
  if (cu) {
    const b = await noiVao(CONNECT_TIMEOUT_MS)
    if (b) return b
    die(`pid ${cu} đang giữ hồ sơ nhưng cổng ${PORT} không trả lời — xem ${LOG}, hoặc {"do":"close"} rồi gọi lại`)
  }
  const bin = BINS.find((b) => existsSync(b))
  if (!bin) die(`không thấy trình duyệt nào ở: ${BINS.join(' · ')}`)
  mkdirSync(PROFILE, { recursive: true })
  mkdirSync(dirname(LOG), { recursive: true })
  // 🔴 KHÔNG `stdio:'ignore'`. Bản đầu vứt hết lời than của chromium, nên lượt
  // hỏng đầu tiên chỉ còn lại một câu "không nối được" — không nói được vì sao.
  // Đúng luật 3 của repo: không đường lỗi nào được đi qua trong im lặng.
  const nhatky = openSync(LOG, 'a')
  const con = spawn(
    bin,
    [
      '--headless=new',
      `--remote-debugging-port=${PORT}`,
      // Chỉ nghe ở máy này: thiếu dòng này thì bất kỳ ai cùng mạng cũng lái
      // được trình duyệt mang phiên đăng nhập của chủ máy.
      '--remote-debugging-address=127.0.0.1',
      `--user-data-dir=${PROFILE}`,
      '--no-first-run',
      '--no-default-browser-check',
    ],
    { detached: true, stdio: ['ignore', nhatky, nhatky] },
  )
  con.unref()
  writeFileSync(LOCK, String(con.pid))
  // Chờ bằng chính phép nối, không bằng một con số `sleep` đoán bừa.
  const han = Date.now() + CONNECT_TIMEOUT_MS
  while (Date.now() < han) {
    const b = await noiVao(1500)
    if (b) return b
    await new Promise((r) => setTimeout(r, 250))
  }
  die(`dựng được trình duyệt (pid ${con.pid}) nhưng cổng ${PORT} không mở sau ${CONNECT_TIMEOUT_MS / 1000}s — xem ${LOG}`)
}

function nhoTrang(url) {
  if (!/^https?:\/\//i.test(url ?? '')) return
  try {
    writeFileSync(LAST, JSON.stringify({ url, at: new Date().toISOString() }))
  } catch {}
}

function trangCuoi() {
  try {
    const u = JSON.parse(readFileSync(LAST, 'utf8')).url
    return /^https?:\/\//i.test(u) ? u : null
  } catch {
    return null
  }
}

async function trang(b) {
  const ctx = b.contexts()[0] ?? (await b.newContext())
  const pages = ctx.pages().filter((p) => !p.isClosed())
  return pages[pages.length - 1] ?? (await ctx.newPage())
}

const lenh = JSON.parse(process.argv[2] ?? '{}')

// `close` chạy được KỂ CẢ khi không nối được: "tắt cái vốn không chạy" là thành
// công, không phải lỗi.
if (lenh.do === 'close') {
  const b = await noiVao(3000)
  if (b) await b.close().catch(() => {})
  const pid = conDangSong()
  if (pid) {
    try {
      process.kill(pid, 'SIGTERM')
    } catch {}
  }
  try {
    writeFileSync(LOCK, '')
  } catch {}
  say({ ok: true, da: b || pid ? 'tat' : 'von khong chay' })
  process.exit(0)
}

let moiDung = false // lượt này có phải tự dựng trình duyệt không
let b = await noiVao(3000)
if (!b) {
  b = await dungTrinhDuyet()
  moiDung = true
}
const p = await trang(b)

// Dựng NGUỘI + trang trắng ⟹ quay về chỗ đang đứng lúc tắt. Không làm với
// `goto` (nó tự đi đâu đó rồi) và không làm khi trang đang có nội dung: người
// ta không muốn một cú nhảy bất ngờ dưới chân mình.
//
// 🔴 "Trang trắng" KHÔNG chỉ là `about:blank` — đo 23/08: hồ sơ đã có lịch sử
// thì Chrome mở thẳng `chrome://new-tab-page/`, nên điều kiện cũ trượt và lượt
// mở lại đứng ở Tab Mới. Bài kiểm chạy thật bắt được ngay; bản đầu đúng chỉ vì
// hồ sơ lúc ấy còn mới tinh.
const trangTrong = (u) => !u || u === 'about:blank' || /^chrome:\/\//i.test(u)
if (moiDung && lenh.do !== 'goto' && trangTrong(p.url())) {
  const cu = trangCuoi()
  if (cu) {
    await p.goto(cu, { waitUntil: 'domcontentloaded', timeout: 30_000 }).catch(() => {})
  }
}

try {
  switch (lenh.do) {
    case 'goto': {
      if (!/^https?:\/\//i.test(lenh.url ?? '')) die('chỉ mở http/https')
      // `domcontentloaded` chứ không `networkidle`: nhiều trang giữ một kết nối
      // mở mãi mãi (websocket, long-poll), nên `networkidle` là chờ tới hết giờ
      // trên đúng những trang đáng xem nhất.
      await p.goto(lenh.url, { waitUntil: 'domcontentloaded', timeout: 30_000 })
      nhoTrang(p.url())
      say({ url: p.url(), title: await p.title() })
      break
    }
    case 'where':
      say({ url: p.url(), title: await p.title() })
      break
    case 'text': {
      const chu = await p.evaluate(() => document.body?.innerText ?? '')
      say({ url: p.url(), title: await p.title(), text: chu })
      break
    }
    case 'shot': {
      if (!lenh.path) die('thiếu `path`')
      await p.screenshot({ path: lenh.path, fullPage: false })
      say({ url: p.url(), title: await p.title(), path: lenh.path })
      break
    }
    // Ba lệnh dưới đây mở ra việc ĐĂNG NHẬP — thứ trước 23/08 không làm được
    // từ xa, và là lý do "đăng nhập một lần còn mãi" mới chỉ chứng minh được
    // tới mức cookie chứ chưa tới mức một phiên thật.
    //
    // ⚠ Chữ gõ ở đây đi QUA TELEGRAM. Với mật khẩu thì đó là một quyết định của
    // chủ máy, không phải của huba — nhưng nó phải được nói ra chứ không lặng
    // lẽ nhận.
    case 'field': {
      if (!lenh.what) die('thiếu `what` — tên ô cần chọn')
      // Tìm theo thứ người đọc NHÌN THẤY, cùng luật với `click`: nhãn, chữ mờ
      // trong ô, rồi mới tới vai trò.
      const o = p
        .getByPlaceholder(lenh.what, { exact: false })
        .or(p.getByLabel(lenh.what, { exact: false }))
        .or(p.getByRole('textbox', { name: lenh.what }))
        .first()
      await o.click({ timeout: 10_000 })
      say({ url: p.url(), title: await p.title(), o: lenh.what })
      break
    }
    case 'type': {
      if (typeof lenh.text !== 'string' || lenh.text === '') die('thiếu `text`')
      // Chưa chọn ô nào thì rơi vào ô nhập ĐẦU TIÊN nhìn thấy được — ngồi ở máy
      // thì người ta bấm vào ô rồi mới gõ, và trang đăng nhập nào cũng để con
      // trỏ sẵn ở ô đầu.
      const dangO = await p.evaluate(() => {
        const e = document.activeElement
        return !!e && (e.tagName === 'INPUT' || e.tagName === 'TEXTAREA' || e.isContentEditable)
      })
      if (!dangO) {
        await p.getByRole('textbox').first().click({ timeout: 10_000 })
      }
      // `delay` để trang nào nghe từng phím (gợi ý, kiểm tra tức thời) kịp chạy.
      await p.keyboard.type(lenh.text, { delay: 20 })
      say({ url: p.url(), title: await p.title(), da_go: lenh.text.length })
      break
    }
    case 'press': {
      const phim = lenh.key || 'Enter'
      const truoc = p.url()
      await p.keyboard.press(phim)
      await p.waitForLoadState('domcontentloaded', { timeout: 15_000 }).catch(() => {})
      // 🔴 `domcontentloaded` KHÔNG bắt được điều hướng của một trang SPA — đo
      // 23/08 trên ô tìm kiếm DuckDuckGo: lệnh trả về ngay, `url` còn là trang
      // cũ và `title` RỖNG, rồi lượt đọc sau mới thấy trang kết quả. Trả một
      // câu như thế là nói rằng cú Enter không làm gì. Nên chờ thêm bằng chính
      // thứ đo được: địa chỉ có đổi không, hoặc tiêu đề đã kịp có chưa.
      for (let i = 0; i < 12 && p.url() === truoc; i++) {
        await new Promise((r) => setTimeout(r, 250))
      }
      for (let i = 0; i < 8 && !(await p.title()); i++) {
        await new Promise((r) => setTimeout(r, 250))
      }
      nhoTrang(p.url())
      say({ url: p.url(), title: await p.title() })
      break
    }
    case 'click': {
      if (!lenh.what) die('thiếu `what`')
      // Tìm theo CHỮ NGƯỜI ĐỌC THẤY, không theo selector: câu lệnh đến từ một
      // người đang nhìn ảnh chụp trang, và thứ họ nhìn thấy là chữ.
      await p.getByText(lenh.what, { exact: false }).first().click({ timeout: 10_000 })
      await p.waitForLoadState('domcontentloaded', { timeout: 15_000 }).catch(() => {})
      nhoTrang(p.url())
      say({ url: p.url(), title: await p.title() })
      break
    }
    default:
      die(`không hiểu lệnh: ${JSON.stringify(lenh.do)}`)
  }
} catch (e) {
  die(String(e?.message ?? e).split('\n')[0])
}
process.exit(0)
